### PR TITLE
5211 change validator field configuration description length limit

### PR DESCRIPTION
**Pull request by IlayDavid95**
This PR adds 2 change validators for the Jira adapter. 1 for maximum description length of a FieldConfiguration (255 characters) and 1 for maximum FieldConfigurationItem description length (1000 characters).

The latter change validator supports both fetch modes for Jira (split fieldconfiguration items and normal)
---
_Release Notes_: 
Jira:
- Added change validator for maximum description length of a FieldConfiguration, limiting to a maximum of 255 characters.
- Added change validator for maximum description length of a FieldConfiguration item, limiting to a maximum of 1000 characters.

---
_User Notifications_: 
Jira: Validation on a FieldConfiguration item description will block deployment if any item has a description with a length exceeding 1000 characters. You might have such description through utilizing the Jira UI.

### CODE REVIEW

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_description_length.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  changes
+    .filter(isInstanceChange)
+    .filter(isModificationChange)
+    .filter(change => getChangeData(change).elemID.typeName === FIELD_CONFIGURATION_TYPE_NAME)
+    .filter(change => change.data.after.value.description.length > FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH)
+    .map(change => ({
+      elemID: getChangeData(change).elemID,
+      severity: 'Error' as SeverityLevel,
+      message: 'Field configuration description length exceeded maximum.',
+      detailedMessage: `Field configuration description length (${change.data.after.value.description.length}) exceeded the allowed maximum of ${FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH} characters.`,
```
Comment (IlayDavid95) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_description_length.ts:39]: Waiting for Lior's approval on this error message.

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        .map(([key, value]: [string, Value]) => ({name: key, description: value.description}))
+        .filter((field: fieldType) => field.description !== undefined && field.description.length > FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH)
+        .map(field => field.name)
+        if (fields.length > 0) { // Should always be true.
+          error.detailedMessage = `Exceeded maximum description length of ${FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH} characters for the following fields: ${fields}.`
+        }
+      }
+
+      return error
+    })
```
Comment (IlayDavid95) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:69]: - [ ] Add empty line at end of file

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { toChange, ObjectType, ElemID, InstanceElement } from '@salto-io/adapter-api'
+import { fieldConfigurationDescriptionLengthValidator } from '../../../src/change_validators/field_configuration/field_configuration_description_length'
+import { JIRA, FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH, FIELD_CONFIGURATION_TYPE_NAME, FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH, FIELD_CONFIGURATION_ITEM_TYPE_NAME} from '../../../src/constants'
+
+describe('fieldConfigurationDescriptionLengthValidator', () => {
```
Comment (IlayDavid95) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts:20]: Separate to different describes.

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      {
+        elemID: fieldConfigurationInstance.elemID,
+        severity: 'Error',
+        message: 'Description length exceeded maximum.',
+        detailedMessage: `Description length (${afterInstance.value.description.length}) exceeded the allowed maximum of ${FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH} characters.`,
+      },
+    ])
+  })
+
+  it.each([0, 1, Math.floor(FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH / 2), FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH])('Should succeed because field configuration item description length is lower than maximum.', async (descriptionLength) => {
```
Comment (IlayDavid95) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts:80]: extract boundary values to const and document.

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_item_description_length.test.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        },
+        item2: {
+          description: 'item2 description'
+        }
+      },
+    })
+
+    fieldConfigurationItemInstance = new InstanceElement('instance', fieldConfigurationItemType, {
+      description: 'item description',
+      id: { elemID: new ElemID(JIRA, 'item') },
```
Comment (IlayDavid95) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_item_description_length.test.ts:48]: Use a reference expression.

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    .map(change => {
+      const error = {
+        elemID: change.elemID,
+        severity: 'Error' as SeverityLevel,
+        message: 'Description length exceeded maximum.',
+        detailedMessage: '',
+      }
+      if (change.elemID.typeName === FIELD_CONFIGURATION_ITEM_TYPE_NAME) {
+        console.log(change)
+        error.detailedMessage = `Description length (${change.value.description.length}) of field configuration item (${change.value.id.elemID.getFullName()}) exceeded the allowed maximum of ${FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH} characters.`
```
Comment (IlayDavid95) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:56]: Waiting for Lior's approval on this error message.

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { toChange, ObjectType, ElemID, InstanceElement } from '@salto-io/adapter-api'
+import { fieldConfigurationDescriptionLengthValidator } from '../../../src/change_validators/field_configuration/field_configuration_description_length'
+import { JIRA, FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH, FIELD_CONFIGURATION_TYPE_NAME } from '../../../src/constants'
+
+describe('fieldConfigurationDescriptionLengthValidator', () => {
+  let fieldConfigurationType: ObjectType
```
Comment (omrifarkash) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts:21]:
```suggestion
  let fieldConfigurationObjectType: ObjectType
```

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  let fieldConfigurationInstance: InstanceElement
+  const boundaryValues = [
+    0, // Lower bound
+    1, // Lower + 1
+    Math.floor(FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH / 2), // Average case
+    FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH - 1, // Upper bound - 1
+    FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH, // Upper bound
```
Comment (omrifarkash) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts:28]: Do we need to test all of these values?

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    ).toEqual([
+      {
+        elemID: fieldConfigurationInstance.elemID,
+        severity: 'Error',
+        message: 'Field configuration description length exceeded maximum.',
+        detailedMessage: `Field configuration description length (${afterInstance.value.description.length}) exceeded the allowed maximum of ${FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH} characters.`,
+      },
+    ])
+  })
+})
```
Comment (omrifarkash) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts:79]: Very nice test!

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_item_description_length.test.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+const boundaryValues = [
+  0, // Lower bound
+  1, // Lower bound + 1
+  Math.floor(FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH / 2), // Average case
+  FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH - 1, // Upper bound - 1
+  FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH, // Upper bound
+]
+
+describe('fieldConfigurationItemDescriptionLengthValidator non split configuration', () => {
+  let fieldConfigurationType: ObjectType
```
Comment (omrifarkash) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_item_description_length.test.ts:34]:
```suggestion
  let fieldConfigurationObjectType: ObjectType
```

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_item_description_length.test.ts [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        elemID: fieldConfigurationInstance.elemID,
+        severity: 'Error',
+        message: 'Description length exceeded maximum.',
+        detailedMessage: `Exceeded maximum description length of ${FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH} characters for the following fields: ${['item1']}.`,
+      },
+    ])
+  })
+})
+
+describe('fieldConfigurationItemDescriptionLengthValidator split item configuration', () => {
```
Comment (omrifarkash) [inline packages/jira-adapter/test/change_validators/field_configuration/field_configuration_item_description_length.test.ts:89]: Please merged the two describe blocks under a single root describe block 'fieldConfigurationItemDescriptionLengthValidator'

You can name the others ['case split item configuration', 'case not split item configuration']

**Review by omrifarkash — COMMENTED** [review]
Really great work!

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_description_length.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+} from '@salto-io/adapter-api'
+import { FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH, FIELD_CONFIGURATION_TYPE_NAME } from '../../constants'
+
+export const fieldConfigurationDescriptionLengthValidator: ChangeValidator = async changes =>
+  changes
+    .filter(isInstanceChange)
+    .filter(isModificationChange)
+    .map(getChangeData)
+    .filter(inst => inst.elemID.typeName === FIELD_CONFIGURATION_TYPE_NAME)
+    .filter(inst => inst.value.description.length > FIELD_CONFIGURATION_DESCRIPTION_MAX_LENGTH)
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_description_length.ts:31]: description is something that will always be define? (O.W this code will throw an error)

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  InstanceElement,
+  ElemID,
+} from '@salto-io/adapter-api'
+import {
+  FIELD_CONFIGURATION_TYPE_NAME,
+  FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH,
+  FIELD_CONFIGURATION_ITEM_TYPE_NAME,
+} from '../../constants'
+
+type describedType = { description: string }
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:32]: We define types in PascalCase
```suggestion
type DescribedType = { description: string }
```

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+type describedElementType = { elemID: ElemID } & describedType
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:33]:
```suggestion
type DescribedElementType = { elemID: ElemID } & describedType
```

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+const isDescriptionTooLong = (obj: describedType): boolean =>
+  obj.description !== undefined && obj.description.length > FIELD_CONFIGURATION_ITEM_DESCRIPTION_MAX_LENGTH
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:35]: I think `obj.description !== undefined` is a redundant check, obj is from type `DescribedType` which must have a description attribute

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+const convertToDescribedElementType = (inst: InstanceElement): describedElementType | describedElementType[] => {
+  if (inst.elemID.typeName === FIELD_CONFIGURATION_ITEM_TYPE_NAME) {
+    return { elemID: inst.elemID, description: inst.value.description }
+  }
+
+  return Object.entries(inst.value.fields).map(([itemName, item]: [string, Value]) => ({
+    elemID: inst.elemID.createNestedID('fields', itemName),
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:42]: You don't need to define the input type inside a map function
```suggestion
  return Object.entries(inst.value.fields).map(([itemName, item]) => ({
    elemID: inst.elemID.createNestedID('fields', itemName),
```

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    description: item.description,
+  }))
+}
+export const fieldConfigurationItemDescriptionLengthValidator: ChangeValidator = async changes =>
+  changes
+    .filter(isInstanceChange)
+    .filter(change => !isRemovalChange(change))
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:49]: We have `isAdditionalOrModificationChange` util
```suggestion
    .filter(isAdditionalOrModificationChange)
```

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    .map(getChangeData)
+    .filter(
+      inst =>
+        inst.elemID.typeName === FIELD_CONFIGURATION_TYPE_NAME ||
+        inst.elemID.typeName === FIELD_CONFIGURATION_ITEM_TYPE_NAME
+    )
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:55]: style nit:
```suggestion
[FIELD_CONFIGURATION_TYPE_NAME, FIELD_CONFIGURATION_ITEM_TYPE_NAME].includes(inst.elemID.typeName)
```

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+    .map(convertToDescribedElementType)
+    .flat()
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:57]: You can use `.flatMap`

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts (hunk 16 above)
Comment (omrifarkash) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:42]: re `inst.elemID.createNestedID('fields', itemName)`
This is the elemID we'd like the changeError to be on?

**Review by omrifarkash — COMMENTED** [review]
Thanks for the new commits.
Please resolve comments that you've fixed and comment on others with explanation why you decided to keep the code as it was :)

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_description_length.ts (hunk 12 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_description_length.ts:31]: Each field configuration is supposed to have a description, AFAIK

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts (hunk 16 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:42]: This change was requested by Lior. He wanted the error to be attached the to field itself.

Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:42]: Also, I get a warning if I remove the typing.

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts (hunk 18 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:55]: This is extremely inefficient, I can use the set again if it is preferred.

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts (hunk 19 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts]: Thanks!

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts (hunk 8 above)
Comment (IlayDavid95) [reply packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts]: I prefer testing multiple values that might behave differently. I removed the MAX_LENGTH - 1 and renamed everything.

File: packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts (hunk 9 above)
Comment (IlayDavid95) [reply packages/jira-adapter/test/change_validators/field_configuration/field_configuration_description_length.test.ts:79]: Thanks.

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts (hunk 15 above)
Comment (omrifarkash) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:38]: Why did you resolve this one?

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts (hunk 16 above)
Comment (omrifarkash) [reply packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:45]: Lets try this f2f

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+export const fieldConfigurationItemDescriptionLengthValidator: ChangeValidator = async changes =>
+  changes
+    .filter(isInstanceChange)
+    .filter(isAdditionOrModificationChange)
+    .map(getChangeData)
+    .filter(
+      inst =>
+        inst.elemID.typeName === FIELD_CONFIGURATION_TYPE_NAME ||
+        inst.elemID.typeName === FIELD_CONFIGURATION_ITEM_TYPE_NAME,
+    )
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:58]: Styling nit

```suggestion
      inst =>
[FIELD_CONFIGURATION_TYPE_NAME, FIELD_CONFIGURATION_ITEM_TYPE_NAME].includes(inst.elemID.typeName)
    )
```

File: packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    .flatMap(convertToDescribedElementType)
+    .filter(isDescribedType)
```
Comment (omrifarkash) [inline packages/jira-adapter/src/change_validators/field_configuration/field_configuration_item_description_length.ts:60]: I don't think you need typeGuard if you already converted the instances to your type

**Review by omrifarkash — APPROVED** [review]
Great
